### DISCUSSION TITLE
TimeLapse WITH upload

### DISCUSSION THREAD

**Discussion by Espusen** [gopro/labs#829] [General]
How is this for a start David?

<img width="1439" alt="Screenshot 2024-03-26 at 20 21 41" src="https://github.com/gopro/labs/assets/153327756/fc9e7841-68fe-4db3-9045-4eb7912230a2">

**Comment by Espusen** [comment]
Nevermind. I see the flaw. I will repost a new one..

**Comment by Espusen** [comment]
This
<img width="1439" alt="Screenshot 2024-03-26 at 20 29 26" src="https://github.com/gopro/labs/assets/153327756/4699e71f-33b9-443b-a2e3-8e63d54f4b63">

**Reply by dnewman-gpsw** [reply]
The third should be >= 16:00 otherwise it will never upload.

**Reply by Espusen** [reply]
Just so I understand the philosophy, why?

**Reply by dnewman-gpsw** [reply]
This is the command you made

```
loop: 
Sleep(300);
if(time > 08:00 AND time < 16:00)
{
  mode(Photo);
  Start();
  Goto loop; 
}
if(time > 17:00)
{
  Join();
  Upload();
}
SleepUntil(08:00);
goto Loop;
```

The error is time can be between 16:00 and 16:69, any of those time, will sleep in 08:00, so this is no change of every seeing a time of > 17:00

**Comment by Espusen** [comment]
Probably need to add some protection lines of code there, like "if no connection with wifi, shutdown after 10 min" or something like that. Testing will reveal what is required but some kind of timeout option would be good I think.

**Comment by Espusen** [comment]
Hi David

I have tried to run the schedule with upload and it seems to sequence correctly but upload is not started. Is this a login issue?


[2024-04-03 17:09:17 !!100%] ACTION:!U!08:00NQ!R
!08:00NQ!R 
[2024-04-03 17:09:17 !!100%] time_max msec:1200000 
[2024-04-03 17:09:17 !!100%] cloud:ready 
[2024-04-03 17:09:20 !!100%] upload waiting:state = 24 
[2024-04-03 17:10:18 !!100%] upload waiting:not triggered 
[2024-04-03 17:10:19 !!100%] upload:not triggered 
[2024-04-03 17:10:19 !!100%] remaining time_max msec:1137900 
[2024-04-03 17:10:19 !!100%] upload:done 
[2024-04-03 17:10:19 !!100%] cloud:still ready 
[2024-04-03 17:10:19 !!100%] ACTION:!08:00NQ!R
!08:00NQ!R 
[2024-04-03 17:10:19 !!100%] Shutdown and Sleep:53376 seconds 
[2024-04-04 07:59:58-delayed ] wake reason:RTC_ALARM 
[2024-04-04 07:59:59 100%  ] Woke with Alarm:[08:00m59]!NQ!R

**Reply by dnewman-gpsw** [reply]
If you press manual upload from the camera's menu, does that work?  Labs uses the same mechanism.

**Comment by Espusen** [comment]
It works fine on manual albeit slow due to poor internet. I’ll report again tomorrow. Very best regardsEspenOn 4 Apr 2024, at 19:25, David Newman ***@***.***> wrote:﻿
If you press manual upload from the camera's menu, does that work?  Labs uses the same mechanism.

**Reply by Espusen** [reply]
It is a funny one this.
Does this log state it is happy with its upload??

[2024-04-05 17:00:47 !!100%] ACTION:!U!08:00NQ!R
!08:00NQ!R 
[2024-04-05 17:00:47 !!100%] time_max msec:1200000 
[2024-04-05 17:00:47 !!100%] cloud:ready 
[2024-04-05 17:00:50 !!100%] upload waiting:state = 24 
[2024-04-05 17:01:48 !!100%] upload waiting:not triggered 
[2024-04-05 17:01:49 !!100%] upload:not triggered 
[2024-04-05 17:01:49 !!100%] remaining time_max msec:1137900 
[2024-04-05 17:01:49 !!100%] upload:done 
[2024-04-05 17:01:49 !!100%] cloud:still ready 
[2024-04-05 17:01:49 !!100%] ACTION:!08:00NQ!R
!08:00NQ!R 
[2024-04-05 17:01:49 !!100%] Shutdown and Sleep:53886 seconds

**Reply by dnewman-gpsw** [reply]
No idea on this.  Labs directly implementing Upload so the logging is not super helpful.  I will need to do more testing as this seems to change with every release.

**Comment by Espusen** [comment]
Hi David

Whats happening here? Battery depletes very rapidly from 14:21 onwards which is a weird time.

https://drive.google.com/file/d/1E_7XCaKv9SdJ6C6WNr6PCbkf7Q0D9kKd/view?usp=drive_link

**Reply by Espusen** [reply]
Sorry David. The link is open now.

**Reply by dnewman-gpsw** [reply]
I had previously discovered this error and it is fixed for the next release, but it is easy to work around.  The issue is the script is not always giving enough time for a file to close before attempting a long sleep. The timing is on the edge, to it doesn't happen every time. In your command you have !S!R!3600N <- this is effective calling order. !R is instant. So you really have !S!3600N and the Photo is taking a second to write to the SD card (so !S is not fully done when it atttempts !3600N, so the camera doesn't sleep, it stays awake for an hour.)   Solution is to use !1N!3599N instead for your sleep, if the previous action is a file close.

**Reply by Espusen** [reply]
Ok Got ya. 
Thank you David! Invaluable!

**Reply by Espusen** [reply]
Is this doable in Blockly or do I change the program after having generated the code?

**Reply by Espusen** [reply]
So this one?
!1N!119N>06:00<20:59mP+!S+!R!06:00NQ!R!08:00NQ!R
It is for a 2 min schedule 0600-2100 with NO upload function
and
!1N!119N>06:00<20:59mP+!S+!R>20:00!W+!U!06:00NQ!R
2 min schedule 0600-2100 WITH upload function

Makes sense? Also Blockly introduces spaces in the code when copied, will that break the chain?

**Reply by dnewman-gpsw** [reply]
Where does blockly add spaces?

Yes they look correct.

**Reply by Espusen** [reply]
If I just copy and paste from Blockly I get this:
!120N>06:00<20:59mP+!S+!R>20:00!W+!U!06:00NQ
!R

The break in the line adds a space it seems.

**Reply by dnewman-gpsw** [reply]
I now understand, just remove the returns.  The returns are the prevent the PNG getting too wide.

**Reply by Espusen** [reply]
Thank you. Just wanted to make sure

**Reply by Espusen** [reply]
Hi David

I have been trying the upload command a few times now and I think the core issue is that it does not allow enough time for the wifi to connect. The camera stays on for about 8-10 seconds. and the message "starting Upload" shows on the front screen for 5-6 seconds, then shuts down. Is there a way to seperate Wifi and Upload commands or otherwise give it longer to do this before it shuts down?
Code: mPdP>06:59<16:59!S+!1N+!299RQ>16:00!W!08:00NQ!3600RQ



How did your film do in Cannes?

**Reply by dnewman-gpsw** [reply]
My film was not official selection, just part of the Short Film Corner market, so it just nice to have a live audience (got good laughs, very satisfying).  

I will explore upload more.

**Reply by Espusen** [reply]
Good to hear! 
Always great to see people value your work

**Comment by dnewman-gpsw** [comment]
I've been testing the upload with this random script:
```mP!S=Ae=A/20=A^2=A+15=A*4!1N<t:N3!U!1N!$AR```

It randomly sets the next interval for a photo, between 1 and 3 minutes.  If the minute on the current time is < 3 it will upload. So this take a photo approximately 30 times an hour, then upload on the hour.  It has been running perfectly for 2 days. 

This is on HERO12. 

Note that I'm only using !U, not !W.  On HERO12 the !U (Upload) will only using the credentials the Quik app provided, this seems reliable. Can you test with only !U?

**Comment by Espusen** [comment]
That sounds promising David!
I will test a different !U script when I get home as I have faster internet there. 
I am currently testing this script: !N!599N>06:00<16:59mP+!S+!R>16:00!W+!U!06:00NQ!R
Not working..

**Reply by dnewman-gpsw** [reply]
The !W is not work or do something unexpected for upload, on HERO12.  My tests never use !W or even setup WiFi credential via Labs.  First use Quik to enable auto off load, but then disable it. So all you are doing is setting up the credentials that manual upload uses.  !U only calls the manual update, same as the camera menu, the camera join the network automatically.  This changed a lot with H12 (base firmware.)

**Reply by Espusen** [reply]
BOOOM!!!!

[2024-06-25 17:06:37 ->--% ] Labs:Initialize 
[2024-06-25 17:06:37 ->--% ] Woke with Command:[17:06m01]!RQ>16:00!U!08:00NQ!3600RQ 
[2024-06-25 17:06:37 ->--% ] Start Now:Past by 37 seconds 
[2024-06-25 17:06:37 ->--% ] Countdown:complete 
[2024-06-25 17:06:37 ->--% ] Timer wake command:!RQ>16:00!U!08:00NQ!3600RQ 
[2024-06-25 17:06:49 !!100%] upload waiting:state = 24 
[2024-06-25 17:07:14 !!100%] upload waiting:enumerating media 
[2024-06-25 17:07:22 !!100%] upload waiting:offloading to sink 
[2024-06-25 17:07:45 !!100%] upload:offloading to sink 
[2024-06-25 17:07:56 !!100%] upload:offloading to sink 
[2024-06-25 17:08:06 !!100%] upload:offloading to sink 
[2024-06-25 17:08:16 !!100%] upload:offloading to sink 
[2024-06-25 17:08:27 !!100%] upload:offloading to sink 
[2024-06-25 17:08:37 !!100%] upload:offloading to sink 
[2024-06-25 17:08:47 !!100%] upload:offloading to sink 
[2024-06-25 17:08:52 !!100%] upload:state = 22 
[2024-06-25 17:08:54 !!100%] upload:done 
[2024-06-25 17:08:55 !!100%] Shutdown and Sleep:53461 seconds 


IT WORKS!!!!
10 Min standard schedule 9-17 daily with upload after last image. 
AMAZING WORK David!!

I will keep it running.

**Reply by Espusen** [reply]
Ok, this may have been a poor restart by me after the initial trial yesterday

I hope the 01-01-00:00 line is not another recurrence of the bug that makes it skip a day?

[2024-06-25 17:06:37 ->--% ] Timer wake command:!RQ>16:00!U!08:00NQ!3600RQ 
[2024-06-25 17:06:49 !!100%] upload waiting:state = 24 
[2024-06-25 17:07:14 !!100%] upload waiting:enumerating media 
[2024-06-25 17:07:22 !!100%] upload waiting:offloading to sink 
[2024-06-25 17:07:45 !!100%] upload:offloading to sink 
[2024-06-25 17:07:56 !!100%] upload:offloading to sink 
[2024-06-25 17:08:06 !!100%] upload:offloading to sink 
[2024-06-25 17:08:16 !!100%] upload:offloading to sink 
[2024-06-25 17:08:27 !!100%] upload:offloading to sink 
[2024-06-25 17:08:37 !!100%] upload:offloading to sink 
[2024-06-25 17:08:47 !!100%] upload:offloading to sink 
[2024-06-25 17:08:52 !!100%] upload:state = 22 
[2024-06-25 17:08:54 !!100%] upload:done 
[2024-06-25 17:08:55 !!100%] Shutdown and Sleep:53461 seconds 
[2024-01-01 00:00:00-delayed ] wake reason:BLE_WAKE 
[2024-06-26 19:08:01 ->--% ] Labs:Display Messages 
[2024-06-26 19:07:53 !!100%] Labs:Main Start 
[2024-06-26 19:07:53 !!100%] Labs:H23.01.02.20.72 
[2024-06-26 19:07:53 !!100%] Labs:Initialize 
[2024-06-26 19:07:53 !!100%] Woke with Command:[08:00m59]!NQ!3600RQ 
[2024-06-26 19:07:56 !!100%] Shutdown and Sleep:46322 seconds

**Reply by Espusen** [reply]
What is BLE_WAKE??

**Reply by dnewman-gpsw** [reply]
BLE event wake the camera, likely some system maintenance when the BLE shutdown (2 hours later seems a clue.)  Labs updated its wake alarm.  That seems all to have worked correctly.

**Reply by Espusen** [reply]
No, the BLE event is 26 hours after the Upload event.
It should have started at 0800 this morning.

**Reply by dnewman-gpsw** [reply]
Yes, I missed the date change. 

This build has more debug https://www.dropbox.com/scl/fi/z81rui7y2xeeytcyjqae2/LABS_HERO12_02_20_72b-sec.zip?rlkey=5h0h2dfnqj2exekuw0jfy9sij&dl=1  and the version I was testing with, just in case I miss something key in the last.

Use *DBGL=1  to see more logging.

**Reply by Espusen** [reply]
I am sad to say that the upload only worked once... 😩
Latest test with the build you provided above..
I am wondering if I shall try again with the build you provided me with before this??
https://drive.google.com/file/d/1-6vKxWyF1mpxiH5gcfmipqE3JN_2DOtX/view?usp=drive_link

**Reply by dnewman-gpsw** [reply]
I still believe that the upload is super network sensitive. It works on my home fiber connected Wi-Fi, 100%, yet switched to a Wi-Fi connected to a 5G access point, fails every time.

**Comment by dnewman-gpsw** [comment]
Time to stop my upload testing, as my account is now complete flooded from working uploads.

**Reply by Espusen** [reply]
Can you get any knowledge out of this?
All it gives is "not triggered"
I beleive it times out way too quick. I am going to see if I can manipulate the router to leave the Port open and see if that makes it respond quicker.

2024-07-03 17:07:17 !!100%] CAMMODE token:P 
[2024-07-03 17:07:17 !!100%] token:P 
[2024-07-03 17:07:18 !!100%] DEFAULT_MODE_OR_DEPTH_OR_DURATION token:P 
[2024-07-03 17:07:18 !!100%] continue from:<16:59!S+!1N+!599RQ>16:00!U!08:00NQ!3600RQ 
[2024-07-03 17:07:18 !!100%] continue from:!S+!1N+!599RQ>16:00!U!08:00NQ!3600RQ 
[2024-07-03 17:07:18 !!100%] continue from:!U!08:00NQ!3600RQ 
[2024-07-03 17:07:18 !!100%] ACTION:!U!08:00NQ!3600RQ 
[2024-07-03 17:07:18 !!100%] time_max msec:1200000 
[2024-07-03 17:07:18 !!100%] cloud:ready 
[2024-07-03 17:07:28 !!100%] upload:not triggered 
[2024-07-03 17:07:28 !!100%] remaining time_max msec:1189900 
[2024-07-03 17:07:28 !!100%] upload:done 
[2024-07-03 17:07:29 !!100%] cloud:still ready 
[2024-07-03 17:07:29 !!100%] ACTION:!08:00NQ!3600RQ 
[2024-07-03 17:07:32 !!100%] Shutdown and Sleep:53546 seconds 
[2024-07-03 17:07:33 !!100%] sleep:53545

**Comment by dbower44022** [comment]
This post has been super helpful, but I have a couple of questions:
1 - What is the purpose of the initial "Sleep 300 Seconds"?  I suppose to create a loop, that slowly checks the "if time > 8:00" statement without executing many times per second?  But, then doesn't the "Sleep until 8:00" stop the loop?  I am having trouble understanding if there are multiple threads or some kind of concurrency...
2 - Does the "Repeat Everything" in the embedded "IF" statement keep the execution until the "time < 16:59" is false?  So the "If time > 17:00" does not execute until after 16:59?
3 - There is no "Stop Capture" before the "Join Network" and "Upload" commands.  Can you upload a file while a "Capture" is still active?
4 - What is the use case for "Sleep Approx seconds" vs "Sleep X Seconds"?  In general why not use absolute values for "Sleep Until"?  Also what is "Sleep and Wait" vs just "Sleep"?

Could you post a image of your completed "Blocky" so I can see what you got to work? (or the command!)

Thanks for any help!  This is incredibly powerful and will help us purchase a bunch of GoPros!

edit:  Here is my program:
![2024-07-01_23-48-25](https://github.com/gopro/labs/assets/36448107/d8a5c3cf-41fc-4bcd-b2df-9cf84ab56538)

**Reply by dnewman-gpsw** [reply]
1. That other loop used Photos, not timelapse, so you will want some quite different. So the 300s was to shutdown the camera to preserve battery between photos. I gather you are on external power.
2. Repeat does repeat everything for the start.  Your code has an error, as the condition > 7am  < 12:25 will be true on every loop (day hours), so it will attempt (but fail) to start the timelapse a thousand times a day.  You need to add a if recording condition.
3. End Capture is the same as stop capture.  In you script you have to end a timelapse, but you don't have to end a single photo (other script).
4.  It is about battery, try ```!60N"Hello"``` and ```!60NQ"Hello"```, time them and watch their behavior. Best to try it, rather than me explaining it.

**Reply by dbower44022** [reply]
Better?

![2024-07-02_09-04-06](https://github.com/gopro/labs/assets/36448107/492b484c-c869-4c71-a20e-b5a3cd219fc6)

**Reply by dbower44022** [reply]
One more question....

If the power is lost before the "End Capture" will the file be saved?
If the script is interrupted before the "End Capture" will the file be saved?  (ex: if I press button?)

Thanks...

**Reply by dnewman-gpsw** [reply]
Yes that is better.  If you have a battery, and the battery is depleting, the file will be closed before shutdown.  If no battery, and power is removed, the file is truncated, and will be repaired on the next boot (but some frames will be lost at the end.)

**Reply by Espusen** [reply]
Just be aware that any timelapse with an interval of a minute or greater is best done as single images with a shutdown in between (as per the code I am experimenting with) and NOT with the timelapse mode which keeps the camera running for, in your case 12 hours!! This is asking for overheat and other issues in my humble opinion. I would certainly not trust it to be left alone for anything longer than one day with your program as it stands now.

**Reply by dbower44022** [reply]
We are using a timelapse with a setting of 2 seconds.  It is in a waterproof enclosure with a battery backup and the battery is connected to a line power charger.  We are certainly asking for overheating, but are looking to figure out some way to reduce heat including:
1 - Use larger enclosure with cooling fan and vent to external air via exhaust and intake pipes in bottom of enclosure.
2 - Use large aluminum enclosure to act as heatsink and mount camera directly to aluminum.
3 - Place shades above enclosure to remove direct sunshine from hitting enclosure.
4 - Modify recording parameters to reduce heat generation
5 - ????

Here is our updated blocky:
![2024-07-05_23-58-37](https://github.com/gopro/labs/assets/36448107/82c13623-ed12-48af-9909-50a6d3afee6b)

**Reply by dnewman-gpsw** [reply]
Has anyone ever experienced overheating a camera in timelapse?  Video and TimeWarp mode are the hot modes, but timelapse is way lower energy.

**Reply by dbower44022** [reply]
The above script was running on my Hero12 Black, with a Sandisk Extreme 512, on USB Power with battery removed, in an airconditioned room with no sunshine and it recorded for about seven hours and then shutdown.  When I powered up....
![IMG_5033](https://github.com/gopro/labs/assets/36448107/06d479ef-207f-4304-8d53-ae3db939684e)



I restarted the script and let it run for a while and then checked the temperature of the front screen and it registered 53C.   I put a small fan (40mm) blowing on the back panel and side and the temperature lowered to 36C and was stable.

I am currently designing a new mount with fans attached to a 40C NO thermistor that I will put in the enclosure to try and keep it cool and allow it to run all day.   The worst part about the over temp failure is that the device does not restart when it finally cools down.  So I will have to go onsite each day, climb up a ladder and restart the power to make sure it did not fail due to overtemp.  :(

**Reply by dnewman-gpsw** [reply]
What was the shooting mode?

**Reply by dbower44022** [reply]
The Blocky used to create the QR code is in my post to this comment above.  The shooting mode was Timelapse, with a frequency of every 2 seconds.  The qr.log file was kind of strange and did not look like a traditional log.  I will post it when I am able to access it next.

**Reply by dnewman-gpsw** [reply]
I find this very strange, useless you had the LCD on all the time, the camera is not working very hard. Plus in an airconditioned room. I've never experienced a thermal shutdown in timelapse, even shooting outdoor in the sun all day, and you aren't using a battery, the main reason for the shutdown. This could be a camera to camera difference.  If you enable debugging, it will log GP2's temperature. 
 
![image](https://github.com/gopro/labs/assets/10856084/ce68972c-4fb6-4cad-97a7-211a3b2a64b2)

**Reply by dnewman-gpsw** [reply]
I run a test shooting timelapse 8:7 5.3k with 2s interval and it run until the external battery ran out, 8 hours later.  The temperature got to a study-state after about an hour, so for 7 hours it was running the GP2 at 67°C, which is low. I then ran the camera at 2.7k240 to heat it up (also at 160Mb/s), to see at what temperature the camera will shutdown (no battery.)  GP2 run until 103°C (20 minutes).  This is why I don't understand your timelapse overheating, timelapse is a low power mode.

**Reply by dbower44022** [reply]
I scanned QR above, and it was recognized by the camera.  I then ran script for 15 minutes and terminated manually.  The log file did not look any different.   Do I need to do something else?  Does it write data in different location?

**Reply by dnewman-gpsw** [reply]
The above is a run time code, not permanent. For permanent *DBGL=1.

**Comment by dbower44022** [comment]
David,
I can't make sense out of it either.  I will install the diagnostics code,
and then try to re-create an over temp failure and hopefully we can get
more information.

When I checked the temperature of my camera using a laser thermometer it
was ~ 53C on the front LCD while running in time lapse mode.  That is lower
than what you were seeing (67C) and nowhere near your test of 103C.  That
is crazy hot!   Did you use my QR Code so we are using the same settings?

I designed and 3d printed a fan mount for the camera and housing, and that
is keeping the temperature much lower (~35C) and my test today ran for over
14 hours while the housing was in the sun with an outside temperature of
26C/80F.  So I think I can get this to work eventually!

I have been having problems with the Sleep Time commands.  But I am
following your other post for that issue, and will update there.

I will keep you informed.

Doug

**Reply by dnewman-gpsw** [reply]
Surface temperature is different than the core processor temperature. The *DBGL=1 will list the core temperature in the log.

**Reply by dbower44022** [reply]
For all following this thread.  We only experienced this issue one time, and have not encountered an overheat since despite running a timelapse video that captures every 2 seconds for 9 months straight.  The overall application has been very reliable as long as the power was connected.  We had one issue where the power was disconnected for over 24 hours and our backup battery drained.  When power was restored the camera began capturing normally.  Great solution for capturing long term construction projects.
